docs/advanced/additional-requests.rst

```diff
@@ -0,0 +1,503 @@
+HTTP downloader implementation in two ways:
+
+.. _setup-contextvars:
+
+1. Context Variable
+-------------------
+
+:mod:`contextvars` is natively supported in :mod:`asyncio` in order to set and
+access context-aware values. This means that the framework using **web-poet**
+can easily assign the implementation using the readily available :mod:`contextvars`
+instance named ``web_poet.request_backend_var``.
+
+This can be set using:
+
+.. code-block:: python
+
+    import attrs
+    import web_poet
+
+    def request_implementation(req: web_poet.HttpRequest) -> web_poet.HttpResponse:
+        ...
+
+
+    def create_http_client():
+        return web_poet.HttpClient()
+
+
+    @attrs.define
+    class SomePage(web_poet.ItemWebPage):
+        http_client: web_poet.HttpClient
+
+        async def to_item(self):
+            ...
+
+    # Once this is set, the ``request_implementation`` will become available to
+    # all instances of HttpClient unless a ``request_downloader`` is injected
+    # to it (see #2 Dependency Injection example below).
+    web_poet.request_backend_var.set(request_implementation)
+
+    # Assume that it's constructed with the necessary arguments taken somewhere.
+    response = web_poet.HttpResponse(...)
+
+    page = SomePage(response=response, http_client=create_http_client())
+    item = page.to_item()
+
+Setting this up would allow access to the request implementation in a
+:class:`~.HttpClient` instance which uses it by default.
+
+.. warning::
+
+    If no value for ``web_poet.request_backend_var`` was set, then a
+    :class:`~.RequestBackendError` is raised. However, no exception would
+    be raised if **option 2** below is used.
+
+
+2. Dependency Injection
+-----------------------
+
+The framework using **web-poet** might be using other libraries which doesn't
+have a full support to :mod:`contextvars` `(e.g. Twisted)`. With that, an
+alternative approach would be to supply the request implementation when creating
+an :class:`~.HttpClient` instance:
+
+.. code-block:: python
+
+    import attrs
+    import web_poet
+
+    def request_implementation(req: web_poet.HttpRequest) -> web_poet.HttpResponse:
+        ...
+
+    def create_http_client():
+        return web_poet.HttpClient(request_downloader=request_implementation)
+
+
+    @attrs.define
+    class SomePage(web_poet.ItemWebPage):
+        http_client: web_poet.HttpClient
+
+        async def to_item(self):
+            ...
+
+    # Assume that it's constructed with the necessary arguments taken somewhere.
+    response = web_poet.HttpResponse(...)
+
+    page = SomePage(response=response, http_client=create_http_client())
+    item = page.to_item()
+
+From the code sample above, we can see that every time an :class:`~.HttpClient`
+is created for Page Objects needing an ``http_client``, the specific **request
+implementation** from a given framework is injected to it.
+
+
+Exception Handling
```

Comment:
I think we should describe error handling before describing backends - just give some examples, of how to handle errors for 

1. individual requests (e.g. get or post)
2. batch requests: what're the semantics, is it possible to get partial data, etc.

There is no need to move the whole section above. I think our docs could have 2 main parts, aimed to a different type of users:

1. People who write Page Objects, and need additional requests
2. People who're implementing web-poet intergations. Information about backends, etc. is mostly for this group.

I think it'd be good to make it clear what to do for (1) first, and then provide information for (2).

Reply:
Gotcha, that makes sense. Thanks for raising it!

I moved the **Exception Handling** section right after discussing how to make additional requests so it's a smoother transition to users developing POs _(as you've pointed out)_. Also written down more examples regarding execption handling in https://github.com/scrapinghub/web-poet/pull/22/commits/7947d588c35edc9e6498480d36281e1d93b8ba2e

Regarding `batch_requests()`, thanks for pointing out about retrieving partial data as I haven't considered it before. The behavior is now updated in https://github.com/scrapinghub/web-poet/pull/22/commits/5d2a7519a0f5b4d0e23e62bd1f2eae099620b37e